src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -56,6 +56,98 @@ DecisionTree<FitnessFunction,
   Train(data, labels, numClasses, minimumLeafSize);
 }
 
+//! Construct and train.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename ElemType,
+         bool NoRecursion>
+template<typename MatType>
+DecisionTree<FitnessFunction,
+             NumericSplitType,
+             CategoricalSplitType,
+             ElemType,
+             NoRecursion>::DecisionTree(MatType&& data,
+                                        const data::DatasetInfo& datasetInfo,
+                                        arma::Row<size_t>&& labels,
+                                        const size_t numClasses,
+                                        const size_t minimumLeafSize)
+{
+  // move the data to avoid copy
+  MatType tmpData = std::move(data);
+  arma::Row<size_t> tmpLabels = std::move(labels);
```

Comment:
I'm reading *Effective Modern C++* today, I think it's a nice solution.

And I think it's a common consideration in mlpack. Whether it's suggested to write a rvalue constructor like this whenever we can't use a single const& without copy?  Maybe this will be added to DesignGuidelines?